src/WebJobs.Script/Description/FunctionInvokerBase.cs

```diff
@@ -81,7 +85,99 @@ protected bool InitializeFileWatcherIfEnabled()
             return false;
         }
 
-        public abstract Task Invoke(object[] parameters);
+        public Task Invoke(object[] parameters)
+        {
+            FunctionStartedEvent startedEvent = null;
+            Task invokeTask = null;
+
+            CancellationToken timeoutOrShutdownToken = (CancellationToken)parameters.First(p => p?.GetType() == typeof(CancellationToken));
+            ExecutionContext functionExecutionContext = (ExecutionContext)parameters.First(p => p?.GetType() == typeof(ExecutionContext));
+            string invocationId = functionExecutionContext.InvocationId.ToString();
+
+            startedEvent = new FunctionStartedEvent(functionExecutionContext.InvocationId, Metadata);
+            _metrics.BeginEvent(startedEvent);
+
+            try
+            {
+                TraceWriter.Info($"Function started (Id={invocationId})");
+
+                invokeTask = InvokeInternal(parameters);
+                invokeTask.Wait(timeoutOrShutdownToken);
+
+                TraceWriter.Info($"Function completed (Success, Id={invocationId})");
+            }
+            catch (OperationCanceledException ex) when (ex.CancellationToken == timeoutOrShutdownToken)
+            {
+                string reason = "Host is Stopping";
+
+                // TODO: What's the best way to do this?
+                var stateProp = typeof(JobHost).GetField("_state", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
+                int hostState = (int)stateProp.GetValue(Host);
+
+                // 3 indicates StoppingOrStopped. If that's the case, just let it go.
+                // Otherwise it's a timeout.
+                if (hostState != 3)
+                {
+                    reason = "Timeout";
+                    ex.Data.Add("FunctionTimeout", true);
```

Comment:
Not doing this anymore...

---
In reply to: [75946188](https://github.com/Azure/azure-webjobs-sdk-script/pull/604#discussion_r75946188) [](ancestors = 75946188)